src/Sentry/Internal/SystemDiagnosticsMetricsListener.cs

```diff
@@ -0,0 +1,87 @@
+#if NET8_0_OR_GREATER
+using System.Diagnostics.Metrics;
+
+namespace Sentry.Internal;
+
+internal class SystemDiagnosticsMetricsListener : IDisposable
+{
+    private readonly Lazy<IMetricAggregator> _metricsAggregator;
+    private IMetricAggregator MetricsAggregator => _metricsAggregator.Value;
+    private static SystemDiagnosticsMetricsListener? DefaultListener;
+
+    internal readonly MeterListener _sentryListener = new ();
+
+    private SystemDiagnosticsMetricsListener(ExperimentalMetricsOptions metricsOptions)
+        : this(metricsOptions, () => SentrySdk.Metrics)
+    {
+    }
+
+    /// <summary>
+    /// Overload for testing purposes - allows us to supply a mock IMetricAggregator
+    /// </summary>
+    internal SystemDiagnosticsMetricsListener(ExperimentalMetricsOptions metricsOptions, Func<IMetricAggregator> metricsAggregatorResolver)
+    {
+        _metricsAggregator = new Lazy<IMetricAggregator>(metricsAggregatorResolver);
+        _sentryListener.InstrumentPublished = (instrument, listener) =>
+        {
+            if (metricsOptions.CaptureSystemDiagnosticsMeters.ContainsMatch(instrument.Meter.Name)
+                || metricsOptions.CaptureSystemDiagnosticsInstruments.ContainsMatch(instrument.Name))
+            {
+                listener.EnableMeasurementEvents(instrument);
+            }
+        };
+        _sentryListener.SetMeasurementEventCallback<byte>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<int>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<float>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<decimal>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<short>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<long>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<double>(RecordMeasurement);
+        _sentryListener.Start();
+    }
+
+    internal static void InitializeDefaultListener(ExperimentalMetricsOptions metricsOptions)
+    {
+        var oldListener = Interlocked.Exchange(
+            ref DefaultListener,
+            new SystemDiagnosticsMetricsListener(metricsOptions)
+            );
+        oldListener?.Dispose();
+    }
+
+    internal void RecordMeasurement<T>(
+        Instrument instrument,
+        T measurement,
+        ReadOnlySpan<KeyValuePair<string, object?>> tags,
+        object? _)
+        where T: struct
+    {
+        var unit = MeasurementUnit.Parse(instrument.Unit);
+        var tagDict = tags.ToImmutableArray().ToImmutableDictionary(
+            kvp => kvp.Key,
+            kvp => kvp.Value?.ToString() ?? string.Empty
+        );
+        var doubleMeasurement = Convert.ToDouble(measurement);
```

Comment:
`where T: struct`

This seems very unsafe. How do we guarantee this won't blow up?
Is it because all data types you set as `T` above are convertable to Double? if so we need some comments here and above to clarify that.
Main concern is someone adding a new value as `T` above and causing issues here

Reply:
Yeah I think something got lost when we made this public (and then reverted to internal). I can add another constraint here `where: struct, IConvertible`. We couldn't do that when it was public becuase `IConvertible` isn't CLS Compliant (can't have it on public interfaces targeting netstandard).